nutils/function.py

```diff
@@ -467,6 +467,279 @@ def lower(self, *, points_shape: Tuple[evaluable.Array, ...] = (), transform_cha
+    :class:`Array` is evaluated to a :class:`numpy.ndarray` with one leading
+    axis compared to the :class:`Array` and all other instances are passed as
+    is. The return value of this method should also include a leading axis with
+    the same length as the other array arguments have, or length one if there
+    are no array arguments. If constructor argument ``npointwise`` is nonzero,
+    the pointwise axes of the :class:`Array` arguments are raveled and included
+    in the single leading axis of the evaluated array arguments as well.
+
+    If possible this method should not use ``self``, e.g. by decorating this
+    method with :func:`staticmethod`. The result of this function must only
+    depend on the arguments and must not mutate the arguments.
+
+    This method is equivalent to ``nutils.evaluable.Array.evalf`` up to
+    the treatment of the leading axis.
+
+    Parameters
+    ----------
+    *args
+        The evaluated arguments corresponding to the ``args`` parameter of the
+        :class:`Custom` constructor.
+
+    Returns
+    -------
+    :class:`numpy.ndarray`
+        The result of this function with one leading points axis.
+    '''
+
+    raise NotImplementedError # pragma: nocover
+
+  def partial_derivative(self, iarg: int, *args: Any) -> IntoArray:
+    '''Return the partial derivative of this function to :class:`Custom` constructor argument number ``iarg``.
+
+    This method is only called for those arguments that are instances of
+    :class:`Array` with dtype :class:`float` and have the derivative target as
+    a dependency. It is therefor allowed to omit an implementation for some or
+    all arguments if the above conditions are not met.
+
+    Axes that are declared pointwise via the ``npointwise`` constructor
+    argument are omitted.
+
+    Parameters
+    ----------
+    iarg : :class:`int`
+        The index of the argument to compute the derivative for.
+    *args
+        The arguments as passed to the constructor of :class:`Custom`.
+
+    Returns
+    -------
+    :class:`Array` or similar
+        The partial derivative of this function to the given argument.
+    '''
+
+    raise NotImplementedError('The partial derivative of {} to argument {} (counting from 0) is not defined.'.format(type(self).__name__, iarg)) # pragma: nocover
+
+class _CustomEvaluable(evaluable.Array):
+
+  def __init__(self, name, evalf, partial_derivative, args: Tuple[Union[evaluable.Array, evaluable.EvaluableConstant], ...], shape: Tuple[int, ...], dtype: DType, points_shape: Tuple[evaluable.Array, ...], transform_chains: Tuple[evaluable.TransformChain, ...], coordinates: Tuple[evaluable.Array, ...]) -> None:
+    self.name = name
+    self.custom_evalf = evalf
+    self.custom_partial_derivative = partial_derivative
+    self.args = args
+    self.points_shape = points_shape
+    self.transform_chains = transform_chains
+    self.coordinates = coordinates
```

Comment:
maybe lump these three in a `lower_args` dict? that shortens what follows somewhat and makes it more future-proof

Reply:
done